File: node-api/src/middleware/jwt.ts
import type { Request, Response, NextFunction } from 'express';
import { verifyToken } from '../services/jwt.js';


/**
 * Middleware JWT para Express.
 * Verifica el token del header Authorization antes de pasar al controller.
 * Si no hay token o es inválido, rechaza con 401.
 */
export function jwtProtected(req: Request, res: Response, next: NextFunction): void {
  const authHeader = req.headers['authorization'];

  if (!authHeader) {
    res.status(401).json({
      status: 401,
      message: 'Se requiere token de autorización. Header: Authorization: Bearer <token>',
    });
    return;
  }

  const parts = authHeader.split(' ');
  if (parts.length !== 2 || parts[0]?.toLowerCase() !== 'bearer') {
    res.status(401).json({
      status: 401,
      message: 'Formato inválido. Usa: Authorization: Bearer <token>',
    });
    return;
  }

  try {
    const payload = verifyToken(parts[1]!);
    // Guardar datos del usuario en el request para los controllers
    (req as Request & { user: typeof payload }).user = payload;
    next();
  } catch {
    res.status(401).json({
      status: 401,
      message: 'Token inválido o expirado',
    });
  }
}

File: go-api/middleware/jwt.go
package middleware

import (
	"interseguro-go-api/services"
	"strings"

	"github.com/gofiber/fiber/v2"
)

// JWTProtected es un middleware que verifica el token JWT en el header Authorization.
// Si el token es válido, guarda los claims en el contexto para que los handlers los usen.
// Si no hay token o es inválido, rechaza con 401.
//
// Uso: app.Use("/api", middleware.JWTProtected())
func JWTProtected() fiber.Handler {
	return func(c *fiber.Ctx) error {
		// Obtener el header Authorization
		authHeader := c.Get("Authorization")
		if authHeader == "" {
			return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
				"status":  401,
				"message": "Se requiere token de autorización. Incluye el header: Authorization: Bearer <token>",
			})
		}

		// El header debe tener el formato "Bearer <token>"
		parts := strings.SplitN(authHeader, " ", 2)
		if len(parts) != 2 || strings.ToLower(parts[0]) != "bearer" {
			return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
				"status":  401,
				"message": "Formato inválido. Usa: Authorization: Bearer <token>",
			})
		}

		tokenStr := parts[1]

		// Validar el token
		claims, err := services.ValidateToken(tokenStr)
		if err != nil {
			return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
				"status":  401,
				"message": "Token inválido o expirado: " + err.Error(),
			})
		}

		// Guardar los datos del usuario en el contexto para los handlers
		c.Locals("username", claims.Username)
		c.Locals("role", claims.Role)

		// Token válido — continuar al siguiente handler
		return c.Next()
	}
}